Closes #1300 NIfTI import

HenkMutsaerts opened the pull request:
### Linked issue

Closes #1300

### How to test

See #1300 and please also check and merge flavor #1300 in the FlavorDatabase.

### Comments

Optional: add helpful comments for the reviewers here

HenkMutsaerts committed:
#1300 xASL_imp_DCM2NII_ConvertScan: bugfix

HenkMutsaerts committed:
#1300 xASL_imp_DCM2NII_Subject_StartConversion: ensure zipping

HenkMutsaerts committed:
#1300 xASL_wrp_NII2BIDS_Subject: T1 vs T1w

HenkMutsaerts committed:
#1300 Only create dir if needed

HenkMutsaerts committed:
#1300 xASL_imp_ReadSourceData

HenkMutsaerts committed:
#1300 xASL_wrp_CreateBiasfield: comment NaN behavior

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_ConvertScan.m:
```diff
@@ -77,16 +79,21 @@
-            bPutInSessionFolder = false;
+            bAnatomical = false;        
+        case {'T1w' 'T2w' 'T1', 'WMH_SEGM', 'FLAIR', 'T2', 'T1c'}
+%             bPutInSessionFolder = false;
```
I don't get this change. bPutInSessionFolder is initialized as true, so if you comment it here, it will put all anat to session folders....

HenkMutsaerts replied:
You need this, otherwise in some cases, import won't work correctly. NIfTI2BIDS (or just before that) the code assumes that even T1's are in their own folders

jan-petr replied:
But previously. This worked rather correctly and you are now changing this behavior. That might be problematic.

HenkMutsaerts replied:
Well let's test it, it doesn't really work nicely, because `NII2BIDS` assumes folders for all scantypes, if I remember correctly. Perhaps especially if you have multiple T1 or multiple FLAIR scans. But there was something going wrong.

jan-petr replied:
We need to check together then.

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_ConvertScan.m:
```diff
@@ -77,16 +79,21 @@
-        case {'T1', 'WMH_SEGM', 'FLAIR', 'T2', 'T1c'}
-            bPutInSessionFolder = false;
+            bAnatomical = false;        
+        case {'T1w' 'T2w' 'T1', 'WMH_SEGM', 'FLAIR', 'T2', 'T1c'}
```
isn't scan_name a keyword? So if you change it here, you might also need to make sure that this is correctly recognized in NII2BIDS. Is there a support for both T1w and T1?

HenkMutsaerts replied:
Yes we probably need to do this. The problem is that it is confusing if we should use T1 or T1w in import, whereas T1w is the BIDS term (and we use both throughout). `scan_name` is not a BIDS keyword, it is something we use in our archaic token structure ;)

jan-petr replied:
Actually, after some though. Yes - OK to use this as this will be renamed to BIDS that has a well defined name for it. 
And I have seen that you have made some further modifications later. 

However, two things have to be done:
1. We need to check how this behaves for T1c. Because previously we had T1w and T1c - which are quite distinct. Now with T1 and T1w, there might be issues that T1c is recognized incorrectly as T1.
2. Functions/BIDS/xASL_bids_Config.m - line 187 and 193 have to be adapted by adding the T1 and T2 types.

HenkMutsaerts replied:
In general: we didn't test it anyway for `T1c`, but we can of course adapt regular expressions to exclude `T1c` when we mean `T1|T1w`.

Here, there is no different behavior for T1|T1w|T1c, this line only says that all structural stuff are `bAnatomical`.

`xASL_bids_Config` line 187 is correct (it has `T1w|T2w|T1c`)

Line 192: can it be `T1(|w)` under filename? This line was already incompatible with line 187, by not having `T1c`?

BTW: my change in `xASL_imp_DCM2NII_ConvertScan` just made sure that both `T1w` and `T1` will be interpreted as anatomical; doesn't necessarily mean that the whole import needs to be able to recognize T1 and T1w (although that would be nice).

Can you fix this further?

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_ConvertScan.m:
```diff
@@ -77,16 +79,21 @@
             bPutInSessionFolder = true;
-        case {'T1', 'WMH_SEGM', 'FLAIR', 'T2', 'T1c'}
-            bPutInSessionFolder = false;
+            bAnatomical = false;        
```
it would be good to initialize this also upfront. not only inside branches

HenkMutsaerts replied:
I actually think nothing should be initialized upfront; those defaults give problems.
Better to provide a default at the end, when an unknown `scan_name` is detected, with a warning. Initializing dependent variables upfront is asking for bugs, unless you assume that we always carefully think of every possible scenario when coding, which is not really possible in our convoluted code anymore ;)

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_ConvertScan.m:
```diff
@@ -162,35 +168,42 @@
-        [TempLine, PrintDICOMFields] = xASL_imp_AppendParmsParameters(parms);
-        summary_line = [summary_line TempLine];
-    else
-        PrintDICOMFields = [];
```
It's wrong to remove PrintDICOMFields from here. And also since this is now inside another IF, you'll need to initialize PrintDICOMFields in the other branch as well.

Because it's an output variable and you will have it missing in certain cases -> error.

HenkMutsaerts replied:
I thought this was never used, this is why I removed it.

jan-petr replied:
We use it - it's the function output. So this has to be reverted.

HenkMutsaerts replied:
Look up to the start of the function, I moved the initialization there, for readability. Just to ensure that we always initialize output parameters at the start.

I came across a debugging issue where I needed that.

jan-petr replied:
OK. Makes sense.

jan-petr commented on Modules/SubModule_Import/xASL_wrp_NII2BIDS_Subject.m:
```diff
@@ -39,7 +39,7 @@
     % by renaming all runs into ASL_1 ASL_2 ASL_n and keeping the
     % unique runs only. Missing scans will issue a warning, not an error.
-    listRuns = xASL_adm_GetFileList(fullfile(x.modules.import.imPar.TempRoot,nameSubjectSession),'^(ASL|T1w|FLAIR).+$',false,[],true);
+    listRuns = xASL_adm_GetFileList(fullfile(x.modules.import.imPar.TempRoot,nameSubjectSession),'^(ASL|T1|T2|FLAIR).+$',false,[],true);
```
are you sure that this is a good idea?

HenkMutsaerts replied:
Well, this way, both T1w and T1 are supported. And T2 as well.
I would think it is the best if the first `dcm2nii` is very flexible, and the subsequent `nii2bids` should be restricted (and we can then solve problems in this stage).

jan-petr replied:
You have convinced me - see the other comment that helps to fix this. Cause we have predefined inputs and assigned outputs. Enough to change this config file to make sure that things are solved on the `nii2bids` side.

HenkMutsaerts replied:
Which other comment?

jan-petr commented on Modules/SubModule_Population/xASL_wrp_CreateBiasfield.m:
```diff
@@ -23,10 +23,27 @@ function xASL_wrp_CreateBiasfield(x)
 %              PM: add normalization of between-subjects SD as well.
```
I'm really confused. Why do we see all the edits here that we've done in a different issue? 
It looks like you have rebase this branch. In general, it is a really really bad idea to rebase a branch anytime before the PR is approved. If necessary to rebase, this has to be done in another way. Cause it makes then really complicated to asses the PR. It's OKish in this case, but otherwise we shouldn't do that?

Or is there another explanation why do I see edits here from another issue?

HenkMutsaerts replied:
No clue!

jan-petr commented on Functions/Import/xASL_imp_DCM2NII_Subject_StartConversion.m:
```diff
@@ -38,18 +38,17 @@
+            % overwrite
+            if x.modules.import.imPar.bOverwrite || ~xASL_exist(first_match,'file') % this will check both .nii & .nii.gz
+                xASL_Copy(scanpath, first_match, x.modules.import.imPar.bOverwrite, x.modules.import.imPar.bVerbose);
+                % gzip if required
```
Original version was unziping, now we are zipping - is that really what we want?

HenkMutsaerts replied:
Yes, we don't want to unzip, when converting to BIDS, we need to zip. So I changed it in a way that `rawdata` will always contains `.nii.gz` (I think a few lines ahead, we run `xASL_adm_GzipAllFiles` to ensure that all are zipped, so this just saves time).

jan-petr reviewed (changes requested):
Changes needed...

HenkMutsaerts commented:
@jan-petr Please go through these changes when looking at the test example that I added to the FlavorDatabase, then you understand some of my suggestions; otherwise, we should go through these together.

HenkMutsaerts committed:
#1300 xASL_imp_NII2BIDS_RunAnat: Allow JSON absence

A NIfTI without JSON for anatomical scans is ideal and not BIDS-compatible, but we can still process, so it should not crash the pipeline (which would only copy 1 anatomical scan and skip any other anatomical scans)

jan-petr commented:
I've resolved some comments and checked new updates - only 3 issues left to be addressed.

HenkMutsaerts committed:
#1300 xASL_bids_BIDSifyM0: minor bugfix

HenkMutsaerts committed:
#1300 xASL_bids_JsonCheck: minor clarification

HenkMutsaerts committed:
#1300 xASL_bids_CompleteBIDS2Legacy: Confusing message

HenkMutsaerts committed:
#1300 xASL_imp_CheckImportSettings: manage (.nii|.nii.gz)

HenkMutsaerts committed:
#1300 xASL_imp_NII2BIDS_Run: avoid creating unneeded folders

HenkMutsaerts commented:
@jan-petr I have resolved the last issues, if you can merge the NIfTI example in the Flavor Database (in branch feature 1300), then this can be closed.

jan-petr committed:
#1300 Fix copying JSONs

jan-petr reviewed (approved):
OK
